components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,470 @@
+        T: Iterator<Item = u8>,
+    {
+        // TODO: make X a usize generic to customize the size of this array
+        // https://github.com/rust-lang/rust/issues/44580
+        // NOTE: 39 is the size required for u128: ceil(log10(u128::MAX)) == 39.
+        const X: usize = 39;
+        // A temporary structure to allow the digits in the iterator to be reversed.
+        // The digits are inserted started from the end, and then a slice is copied
+        // into its final destination (result.digits).
+        let mut mem: [u8; X] = [0u8; X];
+        let mut trailing_zeros: usize = 0;
+        let mut i: usize = 0;
+        for (x, d) in digits_iter.enumerate() {
+            // Take only up to std::i16::MAX values so that we have enough capacity
+            if x > std::i16::MAX as usize {
+                return Err(Error::Limit);
+            }
+            // TODO: Should we check here that `d` is between 0 and 9?
+            // That should always be the case if IntIterator is used.
+            if i != 0 || d != 0 {
+                i += 1;
+                match X.checked_sub(i) {
+                    Some(v) => mem[v] = d,
+                    // This error should be obsolete after X is made generic
+                    None => return Err(Error::Limit),
+                }
+            } else {
+                trailing_zeros += 1;
+            }
+        }
+        let mut result: Self = Default::default();
+        if i != 0 {
+            let magnitude = trailing_zeros + i - 1;
+            debug_assert_le!(magnitude, std::i16::MAX as usize);
+            result.magnitude = magnitude as i16;
+            result.upper_magnitude = result.magnitude;
+            debug_assert_le!(i, X);
+            result.digits.extend_from_slice(&mem[(X - i)..]);
+        }
+        #[cfg(debug_assertions)]
+        result.check_invariants();
+        return Ok(result);
+    }
+
+    /// Gets the digit at the specified order of magnitude. Returns 0 if the magnitude is out of
+    /// range of the currently visible digits.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let dec = FixedDecimal::from(945);
+    /// assert_eq!(0, dec.digit_at(-1));
+    /// assert_eq!(5, dec.digit_at(0));
+    /// assert_eq!(4, dec.digit_at(1));
+    /// assert_eq!(9, dec.digit_at(2));
+    /// assert_eq!(0, dec.digit_at(3));
+    /// ```
+    pub fn digit_at(&self, magnitude: i16) -> u8 {
+        if magnitude > self.magnitude {
+            0 // Leading zero
+        } else {
+            // The following line can't fail: magnitude <= self.magnitude, by
+            // the if statement above, and u16::MAX == i16::MAX - i16::MIN, and
+            // usize is asserted to be at least as big as u16.
+            let j = (self.magnitude as i32 - magnitude as i32) as usize;
+            match self.digits.get(j) {
+                Some(v) => *v,
+                None => 0, // Trailing zero
+            }
```

Comment:
consider `*self.digits.get(j).unwrap_or(&0)`

Reply:
I prefer the way I had written it because it makes it clearer that the `0` refers to a trailing zero.